docs/enhancements/boot-status-reporter.md

```diff
@@ -0,0 +1,129 @@
+---
+title: boot-status-reporter
+authors:
+  - "@nmagnezi"
+creation-date: 2022-09-22
+last-updated: 2022-09-22
+---
+
+# Assisted Boot Reporter
+
+## Summary
+
+Create a new reporting service that runs as early as possible when hosts are booted from disk to collect additional 
+information and send it to `assisted-service`.
+
+## Motivation
+
+There are multiple occurrences of clusters that fail to install at the stage where
+hosts manage to pull ignition but fail to get to a point where the `assisted-controller`
+pod runs and calls home to `assisted-service` with logs.
+
+With that missing information, there's not a whole lot we can do to understand what
+happened with those installations.
+
+
+### Goals
+
+1. Gather information for a host boot stage to improve our ability to debug failures and better inform 
+   users about the host progress stages (first boot, second boot, etc).
+
+
+2. The new service should be thoroughly cleaned up from the hosts and should never fail installations by itself.
+   If the boot reporter service fails to run or report logs, the cluster installation should continue, 
+   and the new host progress stages should be skipped in such a scenario.
+
+## Proposal
+
+
+### Changes To Assisted Service
+For cases as mentioned in `Motivation`, we need a new service (start via `systemd`) soon as possible after hosts
+are booting from the disk and contact `assisted-service` to:
+
+1. Change the installation stage to a new `Booted with local ignition` Between `Configuring`
+   and  `Waiting for control plane`. This stage refers to a boot with pointer ignition.
+
+
+2. Change the installation stage to a new `Booted with control plane ignition` Between
+   `Waiting for control plane` and  `Joined`.
+
+
+3. For each of the above-mentioned, Read the current host stage from `assisted-service`, to determine if:
+   - the service runs on the first boot :arrow_right: collect logs and report.
+   - or the second boot :arrow_right: collect logs, report, and self-cleanup after a configurable threshold.
+
+### Information Needed And Changes To The Discovery Agent
+
+In order to communicate with `assisted-service`, The assisted boot reported service needs to know:
+1. The `assisted-service` IP address and port.
+2. The pull-secret to use for agent authentication.
+3. The InfraEnv ID and host ID
+
+It will need this information starting from the first boot, which is a pointer ignition-based boot.
+
+For that, we need to modify the discovery agent to store this information on disk in a pre-defined location for the assisted boot reporter service to use.
+
+The assisted boot reporter service should start via `systemd` and determine its operating stage. As mentioned above:
+
+### First Boot (Pointer Ignition)
+Expected stage to be read from assisted-service backend: `Configuring`.
```

Comment:
Hmm, I think you are right,
If the host did pull the ignition from the MCS it's already configuring (so what's the point in adding the new stage??). 
If the host didn't pull the ignition from the MCS (failed to reach the API or some other issue with the bootstrap control plane) it will fail during ignition and I doubt if the new boot reporter service will start in that case.

Reply:
>If the host did pull the ignition from the MCS it's already configuring (so what's the point in adding the new stage??).

It's still nice to see a new stage because you know not only it tried to pull ignition but it also began applying it and our service was able to run and our service was able to communicate with the assisted-service, tells you the networking is working great

 >If the host didn't pull the ignition from the MCS (failed to reach the API or some other issue with the bootstrap control plane) it will fail during ignition and I doubt if the new boot reporter service will start in that case.
 
 This is indeed a problem. I wish we could start the new service even when the ignition "remote merge" fails

Reply:
I'm changing it from `Configuring` to `Rebooting` so it will be on-par with the section above.
Also, indeed it does not catch all cases but as Omer said it does let you know that networking is up and running.

Reply:
Done.